### DISCUSSION TITLE
How identify connection between MSC device and PC?

### DISCUSSION THREAD

**Discussion by Stan6314** [hathach/tinyusb#597] [Q&A] [answered]
Hi,
Is there any function that allows identify that the TinyUSB MSC device is connected to a PC?
I would like 2 modes for my device. If it is not connected to a PC, device measures and collects data. As soon as it is connected to the PC, the measurement stops and instrument will be just "USB stick". I would like to test it and decide according to this function whether to measure or not in the main loop.
Thank you for any answer.

**Comment by chegewara** [comment]
How about this:
https://github.com/hathach/tinyusb/blob/master/examples/device/msc_dual_lun/src/main.c#L73-L82

**Comment by hathach** [comment]
`tud_mounted()` as well if you want to poll its status https://github.com/hathach/tinyusb/blob/master/src/device/usbd.h#L63

**Comment by Stan6314** [comment]
Thanks for your answers.
The problem is probably that I use "Adafruit_TinyUSB" fork on Adafruit Feather M0 Express board.
It seems that this library has wrong way for device unmouting. The device remains in mode "is mounted" after disconnection from PC.
Both functions tud_mounted() and tud_umount_cb() behave the same way. The device is still as if connected.
Regards
Stan

**Comment by hathach** [comment]
@Stan6314 for SAMD, you could use the tud_read() https://github.com/hathach/tinyusb/blob/master/src/device/usbd.h#L69 . SAMD controller doesn't distinguish between suspend (standby) and disconnection, which does complicate thing a bit.

**Comment by Stan6314** [comment]
@hathach 
Thanks for your hint to use function tud_ready(). But it was not perfect. Unfortunately, each USB charger (I tested 2 different pcs of 5 V power supply) was detected as a connection to a PC.
Still, it seems that in the end it will work.
I used global variable "bool was_mounted" and combination of functions tud_ready() and tud_mount_cb(). This function is fine unlike tud_umount_cb(). 
In the main loop is "if" statement in the form:

if (was_mounted && (tud_ready()) ) { /* Do nothing here - PC is connected*/ }
else { was_mounted = false; /* and do normal tasks here: measure, display... */  }

In callback function tud_mount_cb(void) just was_mounted variable is set true.
Now it seems (after the first tests) that it could work. This solution recognizes the connection to the PC. And the charger only charges the battery but does not report "PC connection".

Thank you again (twice). First for a great TinyUSB library. Secondly, for helping to solve my problem.

**Comment by hathach** [comment]
I am glad that you figured it out. Though would you mind stating the exact steps that cause the tud_read() failed to detect active connection. It sounds like a bug to me

**Comment by Stan6314** [comment, marked as answer]
Hi hathatch,
it took some time to get to investigate the problem. The whole problem with tud_ready() appears in the sequence of operations. ( I assume that you meant tud_ready() even when you write tud_read()  ).
I used Adafruit Feather M0 Express and Adafruit_TinyUSB_Arduino.
1) When you use tud_ready() and just try to connect the Feather module to the computer, everything works O.K. Every connection or disconnection from the PC is correctly signaled.
2) It's the same even if you use only tud_ready() and connect (after RESET) the charger via the USB connector. Connection to the computer is not reported.
3) But when you connect a computer first and then try to charge the battery, each charger connection is signaled as a connection to a PC. It seems that after the first connection to the PC, something switches internally and then each battery charge is reported as communication with the PC.

The method described above solved this problem. Thank you.
Regards Stan

**Reply by hathach** [reply]
@jcummings2 you could open one, one symptom can be caused by different bug depending on the port.

**Reply by jcummings2** [reply]
Issue #1048 filed with basic sketch file to reproduce the problem.

**Reply by jcummings2** [reply]
For others who come across this thread, here is a better formatted version of the workaround above:

```C++
bool was_mounted = false;
void tud_umount_cb(void) {
    was_mounted  = true;
}

void loop() {
    if (was_mounted && tud_ready()) ) { 
        /* Do nothing here - PC is connected*/
    } else {
        was_mounted = false;
        /* and do normal tasks here: measure, display... */
    }
}
```

Basically, the `was_mounted` variable gets set to `true` on mount and then uses the loop to reset to `false` when disconnected (as determined by `tud_ready()`. A slightly more descriptive variable name might be `mountedSinceLastReady` instead.
